src/main/java/com/google/devtools/build/lib/packages/LabelPrinter.java

```diff
@@ -0,0 +1,104 @@
+package com.google.devtools.build.lib.packages;
+
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.cmdline.PackageIdentifier;
+import com.google.devtools.build.lib.cmdline.RepositoryMapping;
+import com.google.devtools.build.lib.packages.semantics.BuildLanguageOptions;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.StarlarkSemantics;
+
+/** Knows how to print labels consistently in various formats. */
+public interface LabelPrinter {
+  /**
+   * Creates a {@link LabelPrinter} that prints labels in the same way as the Starlark `str` method.
+   * This behavior is useful when matching labels against Starlark values, in particular in tools.
+   *
+   * <p>Do not use this method directly, call {@link
+   * com.google.devtools.build.lib.query2.common.CommonQueryOptions#getLabelPrinter(StarlarkSemantics,
+   * RepositoryMapping)} instead.
+   */
+  static LabelPrinter canonical(StarlarkSemantics starlarkSemantics) {
+    return new LabelPrinter() {
+      @Override
+      public String toString(Label label) {
+        Printer printer = new Printer();
+        label.str(printer, starlarkSemantics);
+        return printer.toString();
+      }
+
+      @Override
+      public String toString(PackageIdentifier packageIdentifier) {
+        // This is adapted from the implementation of Label#str.
```

Comment:
WDYT about just doing `toString(Label.create(packageIdentifier, "DUMMY")).substring(0, -6)`? (imaginary support for "-6" aside)

It avoids the code duplication and makes this much more easily maintainable, at the cost of a not-very-palatable runtime trick (which is presumably not a big deal since how often do we stringify `PackageIdentifier`s?)